fix(TPU): correct dexfee in check balance to prevent swap failures

dimxy opened the pull request:
Two fixes for dexfee to prevent TPU swaps failures:
- correct the platform coin in the params for check balance function (caused ETH swaps failures)
- remove extra dexfee from the taker volume in check balance function (caused TPU swaps with max volume failures)

Also updates TPU swap contracts and ABI byte code to fix TPU docker tests

TODO: add a test to validate spending max volume for TPU

TODO: maybe we should refactor check_balance_for_taker_swap for TPU because it was developed for legacy swaps: assumes legacy dex fee transaction but no TPU funding transaction.

Commit:
fix platform ticker in TakerSwapPreparedParams for taker swap v2

Commit:
update TPU contracts ABI and bytecode

Commit:
improve eth TPU kickstart test to check balances

Commit:
fix(TPU): remove extra dexfee from volume

Commit:
Merge branch 'dev' into fix-TPU-platform-ticker

* dev:
  improvement(`static mut`s): `static mut` removal (#2590)
  fix(orders): set subscription on kickstart and skip GC of own pubkeys (#2597)
  fix(ordermatch): ignore loop-back; clear on null root; reject stale keep-alives (#2580)
  fix(clippy): fix clippy warnings for #2565 (#2589)
  fix(Trezor): fix utxo and eth calls due to firmware changes (#2565)
  fix(utxo): calculate min_trading_vol based on fixed tx fees (#2564)
  feat(protocol): [0] solana support (#2586)
  fix(utxo): fix header deserialization; guard AuxPoW (#2583)
  chore(rust 1.89): make CI clippy/fmt pass (wasm32, all-targets) (#2581)
  fix(utxo): deserialize sapling root for PIVX block headers (#2572)
  improvement(dep-stack): security bumps (#2562)
  fix(utxo): correct block header deserialization for AuxPow and KAWPOW coins (#2563)
  feat(wallet-connect): impl BTC (UTxO) activation via WalletConnect (#2499)
  feat(utxo): add new fixed txfee option for DINGO-like coins (#2454)
  ci(pull-requests): review notification bot (#2468)
  improvement(walletconnect): return the `pairing_topic` in `new_connection` response (#2538)
  bless clippy (#2560)
  refactor(toolchain): use latest available stable compiler (#2557)
  feat(wallet): implement unified offline private key export API (#2542)
  chore(release): v2.3.0-beta (#2284)

# Conflicts:
#	mm2src/mm2_main/tests/docker_tests/eth_docker_tests.rs

Commit:
fix cargo fmt

Commit:
revert erc711 and erc1155 byte code (to use ver with mint fn)

Commit:
revert IERC1155 and IERC721 ABI .json

mariocynicys commented on mm2src/mm2_main/tests/docker_tests/eth_docker_tests.rs:
```diff
@@ -2905,8 +2930,112 @@ fn test_v2_eth_eth_kickstart() {
+                    BigDecimal::from_f64(-0.0001).unwrap(),
+                    BigDecimal::from_f64(0.0001).unwrap(),
+                )
+            };
```
what are the choice of these numbers based on?

dimxy replied:
Those numbers mean that "amount + gas fee" for the swap are okay in some interval.
I did not want to calculate the exact amount (because we have other tests for that purpose)
so I just picked values from this test results and rounded them up and down to create a tolerance interval, to have some rough check.
My idea was to catch hard errors when swaps totally broken and swap produced obviously incorrect amounts.

dimxy replied:
I am open to other suggestions regarding this check, btw

mariocynicys commented on mm2src/mm2_main/tests/docker_tests/eth_docker_tests.rs:
```diff
@@ -2905,8 +2930,112 @@ fn test_v2_eth_eth_kickstart() {
+                .unwrap()
+                .iter()
+                .find(|c| c.get("coin").unwrap().as_str().unwrap() == coin)
+                .unwrap()["protocol"]["type"]
+                .as_str()
+                .unwrap()
+                == "ERC20";
+            let volume = if let Some(price) = price {
+                BigDecimal::from_f64(maker_volume).unwrap() * BigDecimal::from_f64(price).unwrap()
+            } else {
+                BigDecimal::from_f64(maker_volume).unwrap()
+            };
+
+            // add some tolerance
+            let (low_tol, high_tol) = if !is_token {
+                (
+                    // txfee with tolerance
+                    BigDecimal::from_f64(0.0001).unwrap(),
+                    BigDecimal::from_f64(0.0005).unwrap(),
+                )
+            } else {
+                (
+                    // no fee, only tolerance
+                    BigDecimal::from_f64(-0.0001).unwrap(),
+                    BigDecimal::from_f64(0.0001).unwrap(),
+                )
+            };
+
+            if action == "sell" {
+                let low_diff = &volume + &low_tol;
+                let high_diff = &volume + &high_tol;
+                assert!(bal_0 - bal_1 >= low_diff, "{} >= {}", bal_0 - bal_1, low_diff);
+                assert!(bal_0 - bal_1 <= high_diff, "{} <= {}", bal_0 - bal_1, high_diff);
+            } else {
+                let low_diff = &volume - &high_tol;
+                let high_diff = &volume - &low_tol;
+                assert!(bal_1 - bal_0 >= low_diff, "{} >= {}", bal_1 - bal_0, low_diff);
+                assert!(bal_1 - bal_0 <= high_diff, "{} <= {}", bal_1 - bal_0, high_diff);
+            }
+        };
```
let's document why we minus or plus these tolerance values in each case.

dimxy replied:
Refactored this code and added comments: https://github.com/KomodoPlatform/komodo-defi-framework/pull/2600/commits/95acde1cab98d8a5e3f14930815b513c20732fc5

mariocynicys commented on mm2src/mm2_main/src/lp_swap/taker_swap_v2.rs:
```diff
@@ -1029,9 +1029,8 @@ impl<MakerCoin: MmCoin + MakerCoinSwapOpsV2, TakerCoin: MmCoin + TakerCoinSwapOp
+        let preimage_value = TradePreimageValue::Exact(payment_value_with_premium.to_decimal());
         let stage = FeeApproxStage::StartSwap;
 
         let taker_payment_fee = match state_machine
```
i guess we gotta add the dex fee to the argument of `get_sender_trade_fee` for correctness.

dimxy replied:
In fact we should refactor this:
I think, for TPU we should not use this get_sender_trade_fee fn but implement new fee estimation functions corresponding to the TPU steps (send funding and spend funding). 
I though we could do this in a new PR and in this PR just do a quick fix to allow to spend max amount (which actually is needed to run swaps with LR)

shamardy replied:
I agree with your assessment @dimxy regarding needing new fee estimation functions, please note that there was an attempt to add funding spend fees here https://github.com/KomodoPlatform/komodo-defi-framework/pull/2109 so that the taker pays for it. But the PR was never completed due to higher priority items in the past.

shamardy replied:
About `get_sender_trade_fee` we should pass dex fee amount as part of the `preimage_value` since dex fee is paid in the token itself and not the platform coin, it's not a miner / network fee. Now allowance is not correctly calculated if I am not mistaken.

shamardy replied:
Some fixes might be needed in `send_taker_funding_impl` and elsewhere

dimxy replied:
> About `get_sender_trade_fee` we should pass dex fee amount as part of the `preimage_value` since dex fee is paid in the token itself and not the platform coin, it's not a miner / network fee. Now allowance is not correctly calculated if I am not mistaken.

In the get_sender_trade_fee fn we only gas calculate limit for the allowance call but ofc better to pass the correct amount there anyway. 
I guess it's time to add trade fee estimation, specific for TPU.

dimxy replied:
> Some fixes might be needed in `send_taker_funding_impl` and elsewhere

Yes I can see there we do not add dex_fee to the ERC20 payment_amount for handle_allowance but this would not fail the swap as we approve max_value() anyway.

dimxy replied:
I fixed dex_fee in get_sender_trade_fee https://github.com/KomodoPlatform/komodo-defi-framework/pull/2600/commits/b279564bd0eb63a515b4a522dfae5c2503266c6e

I am working on fixes to calculate trade fees specifically for TPU (instead of using legacy get_sender_trade_fee fn) but I will create another PR for that as it's a lot of changes.

mariocynicys reviewed (commented):
Thanks for the fixes!

the last section of `test_v2_eth_eth_kickstart_impl` isn't straight forward to understand. maybe we can add some comments here and there.

Commit:
improve comment

Commit:
fix(review): add dexfee to preimage volume

Commit:
fix(tests): add timeout to make test_approve_erc20 stable

shamardy committed:
Merge remote-tracking branch 'origin/dev' into fix-TPU-dexfee

shamardy commented:
@dimxy both `test_v2_eth_erc20_kickstart` and `test_v2_erc20_eth_kickstart` are failing

Commit:
fix(test): update eth test tolerance (following decreased gas fee due to removed 1 gwei limit in increase_by_percent fn)

Commit:
fix(test): add more gas fee tolerance in balance check

Commit:
refactor(test): improve check tolerance and add more doc comments

dimxy commented:
> @dimxy both `test_v2_eth_erc20_kickstart` and `test_v2_erc20_eth_kickstart` are failing

Fixed the tests (I had to change the check gas fee interval due to the fix in which we removed the 1 gwei gas_price minimum)
